modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java

```diff
@@ -476,4 +554,61 @@ private boolean isMultiPartUpload(String request) {
                 || Regex.simpleMatch("PUT /*/*?*uploadId=*", request);
         }
     }
+
+    public static class TestTelemetryPlugin extends Plugin implements TelemetryPlugin {
+        private static final LongCounter LONG_COUNTER = new LongCounter() {
+            @Override
+            public void increment() {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public void incrementBy(long inc) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public void incrementBy(long inc, Map<String, Object> attributes) {
+                assertThat(
+                    attributes,
+                    allOf(hasEntry("repo_type", S3Repository.TYPE), hasKey("repo_name"), hasKey("operation"), hasKey("purpose"))
+                );
+                METRICS.computeIfAbsent(attributes, k -> new AtomicLong()).addAndGet(inc);
+            }
+
+            @Override
+            public String getName() {
+                return METRIC_REQUESTS_COUNT;
+            }
+        };
+
+        private static final Meter METER = new DelegatingMeter(Meter.NOOP) {
```

Comment:
I also like that we're testing this, but I would prefer we did not use static fields for this. In practice each node will be recording its own metrics, and IMO we should reflect that in these tests. We can do this for instance by tracking these things as instance fields of the plugin - each node has a different plugin instance, which we can get our hands on in the test via `InternalTestCluster.getInstance(PluginsService.class, nodeName).filterPlugins(TestTelemetryPlugin.class)`.

Reply:
Yeah I considered that initially, but decided to go with the static approach mostly because at the end we are interested in aggregated data for dashboards. But I can definitely change it to test each node individually.

Reply:
:+1: yes please, we're breaking new ground in what we're testing here so what we do here will end up establishing the pattern for other similar tests in future.

Reply:
fyi @stu-elastic is working on test framework extensions so that other tests don't have to implement this again.
we will base this on this example or refactor it to use something more generic that would fit other tests too